### PR TITLE
Scala 3 support

### PR DESCRIPTION
**Pull request by oleg-py**
@Asamsig has largely done cross-build setup, I've managed to, for the most part, port the reader/writer and other macros used to Scala 3. VR and native modules still don't compile since they use `@react`. I didn't test those extensively, just made sure it works with existing test suite (wherever `@react` isn't used)

The big sneaky change is that `js.|` is now Scala 3 union type, which behaves very differently (more natural wrt usage and more crazy when it comes to macros and implicit resolution).

There are also a lot of warnings related to machinery used for attribute type-safety, since type projections of form `T#tagType` [are disallowed](https://docs.scala-lang.org/scala3/reference/dropped-features/type-projection.html). The code _currently compiles_ with `-source:3.0-migration` flag though, but I'm unsure if that's going to actually _work_ in user-land (and that flag is really bad for some Scala 3 features and prevents some new syntax from working, so slinky users wouldn't want that).

Not too sure what the plans are for `@react`-style DSL. It would definitely be a huge pain for me to write code like `Input.component(Input.Props(...))`. I've been looking into making that convenient, and while you can't synthesize members, it's potentially possible to use `Dynamic` with a macro. I, however, was unable to figure out a macro rewrite from `Input(...)` to that above construct that would support named parameters.

### CODE REVIEW

**Comment by Asamsig** [comment]
Very impressive, great to see that you were able to get support for recursive case classes, value classes and default values, is there any outstanding issues then for read/write? I guess opaque types and Scala 3 enums might not work yet, but most important thing now is parity with Scala 2, in my opinion at least.

`-source:3.0-migration` should only be temporary, we should try to migrate all incompatible code to Scala 3, would be great with a MVP though, since I imagine it might be quite a chore.

**Comment by shadaj** [comment]
This is absolutely incredible! Am wrapping up my summer internship this/next week, but will try to look through the details this weekend!

**Comment by oleg-py** [comment]
@Asamsig 
> is there any outstanding issues then for read/write? 
Not really on a surface level, but I specifically haven't done any checking for complex cases like classes with multiple constructors, type parameters, etc.

> I guess opaque types and Scala 3 enums might not work yet
Never checked those, but enums should be able to work with mirrors (well, maybe not GADTs). Opaque types might be workable with the same approach I used for value types (custom mirror-like type providing conversion/type info with a macro implicit), but they would all "work" anyway with a fallback reader/writer.

@shadaj thanks, and good luck!

File: build.sbt [hunk 1]
```diff
@@ -3,17 +3,18 @@ ThisBuild / organization := "me.shadaj"
...
 )
 
-val scala212 = "2.12.10"
-val scala213 = "2.13.2"
+val scala212 = "2.12.14"
+val scala213 = "2.13.6"
+val scala3   = "3.0.1"
 
-ThisBuild / scalaVersion := scala213
+ThisBuild / scalaVersion := scala3
```
Comment (shadaj) [inline build.sbt:16]: We'll need to keep the default version on `scala213` to get style checks to pass.

File: tests/src/test/scala/slinky/core/annotations/LocalImportsComponent.scala [hunk 2]
```diff
@@ -1,11 +1,11 @@
 package slinky.core.annotations
 
-// compile-only test, this used to crash the macro annotation
-object LocalImportsComponent {
-  import slinky.core.StatelessComponent
+// // compile-only test, this used to crash the macro annotation
```
Comment (shadaj) [inline tests/src/test/scala/slinky/core/annotations/LocalImportsComponent.scala:3]: Could we move this (and the other commented out tests) to a `scala-2` source folder until we can handle `@react` properly?

**Comment by shadaj** [comment]
I think for the first pass, we should ignore `@react` and just get the basics working. for `T#tagType`, I wonder if we could just use regular dependent type syntax and refer to `attr.tagType`? I remember that we had to use `T#tagType` to get around some issues with dependent types in Scala 2.

**Comment by oleg-py** [comment]
Looks like Scala.js is only available on 2.12.14 since 1.5.1, and I'm not sure what's the lowest version that targets Scala 3. There's no 0.6.x for the most modern versions of Scala 2.12, 2.13 and 3.x either. I can downgrade Scala-2 versions back to where 0.6 existed and skip building for 3 if ScalaJS is not 1.x, though not sure if maintaining the 0.6 build is worth the trouble.

No idea what to do with dependent types yet - the thing is, we need to have a value (maybe an erased one would work) to project the type out. So the `type supports[T <: Tag] = AttrPair[attrType] => AttrPair[T#tagType]` won't work. Maybe something in the new features like polymorhpic values and/or context functions that could replace it though

**Comment by shadaj** [comment]
@oleg-py I'd be fine with dropping 0.6 builds, since this PR will be part of the 0.7.0 release and by now hopefully most users will have moved over to Scala.js 1.0. For `type supports`, I wonder if we could use the type function pattern matching features to extract out the `tagType` (just throwing out an idea, haven't written too much Scala 3 code myself yet). This doesn't seem like a use case that's too unreasonable for Scala 3.

**Comment by shadaj** [comment]
@oleg-py it seems like there's a fair bit of demand for partial Scala 3 support, so let's see if we can get this merged! I think the current set of supported features should be fine (`supports` and `@react` have sufficient workarounds).

Looks like right now CI is failing due to style checks?

**Comment by shadaj** [comment]
Looks like `native` is failing (and `vr` will probably fail too). Let's disable those on Scala 3 for now, and just get `core` and `web` working since that's what most folks depend on.

**Review by shadaj — APPROVED** [review]
LGTM! Thanks for the amazing work @oleg-py and @Asamsig!

**Pull request merged**